# <font color='blue'>Data Science Academy - Python Fundamentos - Capítulo 7</font>

## Download: http://github.com/dsacademybr

In [26]:
# Versão da Linguagem Python
from platform import python_version
print('Versão da Linguagem Python Usada Neste Jupyter Notebook:', python_version())

Versão da Linguagem Python Usada Neste Jupyter Notebook: 3.7.6


## Missão: Analisar o Comportamento de Compra de Consumidores.

## Nível de Dificuldade: Alto

Você recebeu a tarefa de analisar os dados de compras de um web site! Os dados estão no formato JSON e disponíveis junto com este notebook.

No site, cada usuário efetua login usando sua conta pessoal e pode adquirir produtos à medida que navega pela lista de produtos oferecidos. Cada produto possui um valor de venda. Dados de idade e sexo de cada usuário foram coletados e estão fornecidos no arquivo JSON.

Seu trabalho é entregar uma análise de comportamento de compra dos consumidores. Esse é um tipo de atividade comum realizado por Cientistas de Dados e o resultado deste trabalho pode ser usado, por exemplo, para alimentar um modelo de Machine Learning e fazer previsões sobre comportamentos futuros.

Mas nesta missão você vai analisar o comportamento de compra dos consumidores usando o pacote Pandas da linguagem Python e seu relatório final deve incluir cada um dos seguintes itens:

** Contagem de Consumidores **

* Número total de consumidores


** Análise Geral de Compras **

* Número de itens exclusivos
* Preço médio de compra
* Número total de compras
* Rendimento total


** Informações Demográficas Por Gênero **

* Porcentagem e contagem de compradores masculinos
* Porcentagem e contagem de compradores do sexo feminino
* Porcentagem e contagem de outros / não divulgados


** Análise de Compras Por Gênero **

* Número de compras
* Preço médio de compra
* Valor Total de Compra
* Compras for faixa etária


** Identifique os 5 principais compradores pelo valor total de compra e, em seguida, liste (em uma tabela): **

* Login
* Número de compras
* Preço médio de compra
* Valor Total de Compra
* Itens mais populares


** Identifique os 5 itens mais populares por contagem de compras e, em seguida, liste (em uma tabela): **

* ID do item
* Nome do item
* Número de compras
* Preço do item
* Valor Total de Compra
* Itens mais lucrativos


** Identifique os 5 itens mais lucrativos pelo valor total de compra e, em seguida, liste (em uma tabela): **

* ID do item
* Nome do item
* Número de compras
* Preço do item
* Valor Total de Compra


** Como considerações finais: **

* Seu script deve funcionar para o conjunto de dados fornecido.
* Você deve usar a Biblioteca Pandas e o Jupyter Notebook.


In [27]:
# Imports
import pandas as pd
import numpy as np

In [28]:
# Carrega o arquivo
load_file = "dados_compras.json"
df = pd.read_json(load_file, orient = "records")
df.head()

Login  Idade       Sexo  Item ID  \
0     Aelalis34     38  Masculino      165   
1        Eolo46     21  Masculino      119   
2   Assastnya25     34  Masculino      174   
3  Pheusrical25     21  Masculino       92   
4        Aela59     23  Masculino       63   

                                Nome do Item  Valor  
0                Bone Crushing Silver Skewer   3.37  
1  Stormbringer, Dark Blade of Ending Misery   2.32  
2                            Primitive Blade   2.46  
3                               Final Critic   1.36  
4                             Stormfury Mace   1.27

## Informações Sobre os Consumidores

In [29]:
df

Login  Idade       Sexo  Item ID  \
0       Aelalis34     38  Masculino      165   
1          Eolo46     21  Masculino      119   
2     Assastnya25     34  Masculino      174   
3    Pheusrical25     21  Masculino       92   
4          Aela59     23  Masculino       63   
..            ...    ...        ...      ...   
775       Eural50     22  Masculino       98   
776    Lirtossa78     14  Masculino      104   
777    Tillyrin30     20  Masculino      117   
778    Quelaton80     20  Masculino       75   
779        Alim85     23   Feminino      107   

                                  Nome do Item  Valor  
0                  Bone Crushing Silver Skewer   3.37  
1    Stormbringer, Dark Blade of Ending Misery   2.32  
2                              Primitive Blade   2.46  
3                                 Final Critic   1.36  
4                               Stormfury Mace   1.27  
..                                         ...    ...  
775                Deadline, Voice Of Subtlety   3.62  
776                         Gladiator's Glaive   1.36  
777          Heartstriker, Legacy of the Light   4.15  
778                    Brutality Ivory Warmace   1.72  
779                  Splitter, Foe Of Subtlety   3.61  

[780 rows x 6 columns]

In [30]:
len(df['Login'].unique())

573

### O WebSite teve um total de 573 logins diferentes realizando compras

## Análise Geral de Compras
- Número de itens exclusivos
- Preço médio de compra
- Número total de compras
- Rendimento total

### Os seguintes itens foram vendidos

In [31]:
for prod in df['Nome do Item'].unique():
    print(prod)

Bone Crushing Silver Skewer
Stormbringer, Dark Blade of Ending Misery
Primitive Blade
Final Critic
Stormfury Mace
Sleepwalker
Mercenary Sabre
Interrogator, Blood Blade of the Queen
Ghost Reaver, Longsword of Magic
Expiration, Warscythe Of Lost Worlds
Despair, Favor of Due Diligence
Alpha, Reach of Ending Hope
Dreamkiss
Piety, Guardian of Riddles
Bonecarvin Battle Axe
Blood-Forged Skeletal Spine
Twilight's Carver
Lightning, Etcher of the King
Celeste
Winterthorn, Defender of Shifting Worlds
Glimmer, Ender of the Moon
Phantomlight
Brimstone
Dragon's Greatsword
Conqueror Adamantite Mace
Persuasion
Crying Steel Sickle
The Oculus, Token of Lost Worlds
Glinting Glass Edge
War-Forged Gold Deflector
Shadow Strike, Glory of Ending Hope
Rage, Legacy of the Lone Victor
Betrayer
Serenity
Scalpel
Hero Cane
Thorn, Satchel of Dark Souls
Venom Claymore
Storm-Weaver, Slayer of Inception
Malificent Bag
Agatha
Lazarus, Terror of the Earth
Ritual Mace
Severance
Orenmir
Endbringer
Striker
Trickster
Retribu

### O número total de compras foi 780, cada uma tendo valor médio de 2.93

In [32]:
df.describe()

Idade     Item ID       Valor
count  780.000000  780.000000  780.000000
mean    22.729487   91.292308    2.931192
std      6.930604   52.707897    1.115780
min      7.000000    0.000000    1.030000
25%     19.000000   44.000000    1.960000
50%     22.000000   91.000000    2.880000
75%     25.000000  135.000000    3.910000
max     45.000000  183.000000    4.950000

In [33]:
df['Valor'].sum()

2286.33

### Totalizando uma renda total de 2286,33

## Análise Demográfica

## Informações Demográficas Por Gênero
- Porcentagem e contagem de compradores masculinos
- Porcentagem e contagem de compradores do sexo feminino
- Porcentagem e contagem de outros / não divulgados

In [34]:
dfLoginSex = df[['Login', 'Sexo']].drop_duplicates()
dfLoginSex['Sexo'].value_counts()
total = 465 + 100 + 8
Masc = 465
percMasc = 100*Masc / total
Fem = 100
percFem = 100 * Fem / total
outro = 8
percOutro = 100 * outro / total

In [35]:
print(f'Foram um total de {total} consumidores, sendo: \n \
- {Masc} do sexo Masculino ({percMasc:.2f}%) \n \
- {Fem} do sexo Feminino ({percFem:.2f}%) \n \
- {outro} do sexo outro/não divulgado ({percOutro:.2f}%)')

Foram um total de 573 consumidores, sendo: 
 - 465 do sexo Masculino (81.15%) 
 - 100 do sexo Feminino (17.45%) 
 - 8 do sexo outro/não divulgado (1.40%)


## Análise de Compras Por Gênero
- Número de compras
- Preço médio de compra
- Valor Total de Compra
- Compras for faixa etária

In [48]:
df['FaixaEtaria'] = ''
for i, v in enumerate(df['Idade']):
    if v < 18:
        df['FaixaEtaria'][i] = 'Até 18 anos'
    elif 18 <= v <= 25:
        df['FaixaEtaria'][i] = '18 - 25 anos'
    elif 25 < v <= 40:
        df['FaixaEtaria'][i] = '26 - 40 anos'
    elif 40 < v <= 60:
        df['FaixaEtaria'][i] = '41 - 60 anos'
    elif v > 60:
        df['FaixaEtaria'][i] = 'Mais de 60 anos'

In [49]:
df_sexGroup = df[['Idade', 'Valor', 'FaixaEtaria']].groupby(by = df['Sexo'])
df_sexGroup.describe()

Idade                                               \
                       count       mean       std   min   25%   50%   75%   
Sexo                                                                        
Feminino               136.0  22.558824  7.419459   7.0  18.0  22.0  25.0   
Masculino              633.0  22.685624  6.804740   7.0  20.0  22.0  25.0   
Outro / Não Divulgado   11.0  27.363636  6.932139  12.0  24.0  27.0  33.0   

                             Valor                                           \
                        max  count      mean       std   min     25%    50%   
Sexo                                                                          
Feminino               40.0  136.0  2.815515  1.151027  1.03  1.8275  2.615   
Masculino              45.0  633.0  2.950521  1.109967  1.03  2.0400  2.910   
Outro / Não Divulgado  34.0   11.0  3.249091  0.957230  1.88  2.2850  3.730   

                                    
                         75%   max  
Sexo                                
Feminino               3.750  4.95  
Masculino              3.910  4.95  
Outro / Não Divulgado  3.985  4.32

In [50]:
df_sexGroup['Valor'].sum()

Sexo
Feminino                  382.91
Masculino                1867.68
Outro / Não Divulgado      35.74
Name: Valor, dtype: float64

In [51]:
df_sexGroup['FaixaEtaria'].value_counts()

Sexo                   FaixaEtaria 
Feminino               18 - 25 anos     73
                       26 - 40 anos     33
                       Até 18 anos      30
Masculino              18 - 25 anos    374
                       26 - 40 anos    137
                       Até 18 anos     119
                       41 - 60 anos      3
Outro / Não Divulgado  26 - 40 anos      8
                       18 - 25 anos      2
                       Até 18 anos       1
Name: FaixaEtaria, dtype: int64

In [52]:
df_FaixaEtariaGroup = df[['Idade', 'Valor', 'Sexo']].groupby(by=df['FaixaEtaria'])

In [53]:
sexo = df['Sexo'].unique()
df_FaixaMasc = df[['Idade', 'Valor', 'Sexo']].query(f"Sexo == '{sexo[0]}'").groupby(by=df['FaixaEtaria'])
df_FaixaFem = df[['Idade', 'Valor', 'Sexo']].query(f"Sexo == '{sexo[1]}'").groupby(by=df['FaixaEtaria'])
df_FaixaOutro = df[['Idade', 'Valor', 'Sexo']].query(f"Sexo == '{sexo[2]}'").groupby(by=df['FaixaEtaria'])

In [58]:
df_FaixaMasc.describe().round(2)

Idade                                             Valor        \
              count   mean   std   min   25%   50%   75%   max  count  mean   
FaixaEtaria                                                                   
18 - 25 anos  374.0  21.99  2.06  18.0  20.0  22.0  24.0  25.0  374.0  2.96   
26 - 40 anos  137.0  32.20  4.14  26.0  29.0  31.0  35.0  40.0  137.0  2.97   
41 - 60 anos    3.0  43.33  1.53  42.0  42.5  43.0  44.0  45.0    3.0  2.88   
Até 18 anos   119.0  13.39  3.20   7.0  11.0  15.0  15.0  17.0  119.0  2.89   

                                                  
               std   min   25%   50%   75%   max  
FaixaEtaria                                       
18 - 25 anos  1.11  1.03  2.17  2.94  3.91  4.95  
26 - 40 anos  1.12  1.03  2.04  2.86  3.96  4.95  
41 - 60 anos  0.86  2.11  2.42  2.72  3.26  3.81  
Até 18 anos   1.12  1.03  1.94  2.88  3.82  4.95

In [57]:
df_FaixaFem.describe().round(2)

Idade                                             Valor        \
             count   mean   std   min    25%   50%   75%   max count  mean   
FaixaEtaria                                                                  
18 - 25 anos  73.0  22.03  2.29  18.0  20.00  22.0  24.0  25.0  73.0  2.80   
26 - 40 anos  33.0  32.33  5.06  26.0  27.00  32.0  37.0  40.0  33.0  2.83   
Até 18 anos   30.0  13.10  3.61   7.0  10.25  14.0  16.0  17.0  30.0  2.83   

                                                  
               std   min   25%   50%   75%   max  
FaixaEtaria                                       
18 - 25 anos  1.24  1.03  1.67  2.52  3.91  4.95  
26 - 40 anos  1.04  1.03  2.07  2.88  3.57  4.95  
Até 18 anos   1.07  1.27  1.90  2.54  3.68  4.89

In [56]:
df_FaixaOutro.describe().round(2)

Idade                                              Valor        \
             count   mean   std   min    25%   50%    75%   max count  mean   
FaixaEtaria                                                                   
18 - 25 anos   2.0  21.50  0.71  21.0  21.25  21.5  21.75  22.0   2.0  3.30   
26 - 40 anos   8.0  30.75  3.45  26.0  27.00  32.5  33.25  34.0   8.0  3.14   
Até 18 anos    1.0  12.00   NaN  12.0  12.00  12.0  12.00  12.0   1.0  4.00   

                                                 
               std   min   25%  50%   75%   max  
FaixaEtaria                                      
18 - 25 anos  1.34  2.36  2.83  3.3  3.78  4.25  
26 - 40 anos  0.98  1.88  2.15  3.5  3.84  4.32  
Até 18 anos    NaN  4.00  4.00  4.0  4.00  4.00

### Total de compras:
- Feminno: 136
- Masculino: 633
- Outro / Não Divulgado: 11

### Preço Médio de compras
- Feminino: 2.81
- Masculino: 2.95
- Outro / Não Divulgado: 3.25

### Valor Total
- Feminino: 382.91
- Masculino: 1867.68
- Outro / Não Divulgado: 35.74

### Perfil de comprar por faixa etária (Média do valor)
Feminino
- Até 18 anos: 2.83
- 18 - 25 anos: 2.80
- 26 - 40 anos: 2.83
- 41 - 60 anos: Nenhum
- Mais de 60 anos: Nenhum

Masculino
- Até 18 anos: 2.89
- 18 - 25 anos: 2.96
- 26 - 40 anos: 2.97
- 41 - 60 anos: 2.88
- Mais de 60 anos: Nenhum

Outro / Não Divulgado
- Até 18 anos: 4.00
- 18 - 25 anos: 3.30
- 26 - 40 anos: 3.15
- 41 - 60 anos: Nenhum
- Mais de 60 anos: Nenhum

## Consumidores Mais Populares (Top 5)
- Login
- Número de compras
- Preço médio de compra
- Valor Total de Compra
- Itens mais populares

### Top 5 - Login com número de compras

In [66]:
df['Login'].value_counts().sort_values(ascending=False).head(5)

Undirrala66    5
Qarwen67       4
Mindimnya67    4
Hailaphos89    4
Sondastan54    4
Name: Login, dtype: int64

In [108]:
top5Consumers = []
for k, v in df['Login'].value_counts().sort_values(ascending=False).head().items():
    top5Consumers.append(k)

print('Valor médio de compra dos 5 principais consumidores (com base no número de compras)')
df[['Login','Valor']].query(f'Login in {top5Consumers}').sort_values(by='Login').groupby(by = 'Login').mean()

Valor médio de compra dos 5 principais consumidores (com base no número de compras)


Valor
Login              
Hailaphos89  1.4675
Mindimnya67  3.1850
Qarwen67     2.4925
Sondastan54  2.5600
Undirrala66  3.4120

In [109]:
print('Valor total de compra dos 5 principais consumidores (com base no número de compras)')

df[['Login', 'Valor']].query(f'Login in {top5Consumers}').sort_values(by='Login').groupby(by='Login').sum()

Valor total de compra dos 5 principais consumidores (com base no número de compras)


Valor
Login             
Hailaphos89   5.87
Mindimnya67  12.74
Qarwen67      9.97
Sondastan54  10.24
Undirrala66  17.06

In [114]:
print('Os 5 produtos mais comprados')
df['Nome do Item'].value_counts().sort_values(ascending=False).head()

Os 5 produtos mais comprados


Final Critic                            14
Arcane Gem                              11
Betrayal, Whisper of Grieving Widows    11
Stormcaller                             10
Retribution Axe                          9
Name: Nome do Item, dtype: int64

## Itens Mais Populares
- ID do item
- Nome do item
- Número de compras
- Preço do item
- Valor Total de Compra
- Itens mais lucrativos

In [127]:
df = df[['Login', 'Idade', 'Sexo', 'Item ID', 'Item', 'Valor', 'FaixaEtaria']]


In [139]:
itensMaisPopulares = []
for k, v in df['Item'].value_counts().sort_values(ascending=False).head().items():
    itensMaisPopulares.append(k)

dfTopItens = df.query(f'Item in {itensMaisPopulares}').sort_values(by = 'Item')
dfTopItens[['Item ID', 'Item']].groupby(by = 'Item ID').describe()

Item                                                  
        count unique                                   top freq
Item ID                                                        
30          5      1                           Stormcaller    5
34          9      1                       Retribution Axe    9
39         11      1  Betrayal, Whisper of Grieving Widows   11
84         11      1                            Arcane Gem   11
92          8      1                          Final Critic    8
101         6      1                          Final Critic    6
180         5      1                           Stormcaller    5

- O DataFrame acima mostra os protudos mais comprados, com ItemID e a quantidade de vezes em que foram comprados.
- Inclusive pode-se perceber um erro em que um produto (Final Critic) possui dois IDs diferentes.

In [146]:
dfTopItens[['Item', 'Valor']].groupby(by='Valor').describe()

Item                                                  
      count unique                                   top freq
Valor                                                        
1.36      8      1                          Final Critic    8
2.23     11      1                            Arcane Gem   11
2.35     11      1  Betrayal, Whisper of Grieving Widows   11
2.78      5      1                           Stormcaller    5
4.14      9      1                       Retribution Axe    9
4.15      5      1                           Stormcaller    5
4.62      6      1                          Final Critic    6

- O DataFrame acima mostra os principais produtos com os valores.
- Pode-se observar que também existem diferenças nos valores, os produtos StormCaller e Final Critic possuem dois valores diferentes

In [148]:
dfTopItens[['Item', 'Valor']].groupby(by='Item').sum()

Valor
Item                                       
Arcane Gem                            24.53
Betrayal, Whisper of Grieving Widows  25.85
Final Critic                          38.60
Retribution Axe                       37.26
Stormcaller                           34.65

- Com o DataFrame acima pode-se observar o valor total vendido dos 5 produtos mais vendidos

In [152]:
df[['Item', 'Valor']].groupby(by='Item').sum().sort_values(by = 'Valor', ascending=False).head()

Valor
Item                             
Final Critic                38.60
Retribution Axe             37.26
Stormcaller                 34.65
Spectral Diamond Doomblade  29.75
Orenmir                     29.70

## Itens Mais Lucrativos
- ID do item
- Nome do item
- Número de compras
- Preço do item
- Valor Total de Compra

In [173]:
prodMaisLucrativos = []
for k,v in df[['Item', 'Valor']].groupby(by='Item').sum().sort_values(by = 'Valor', ascending=False).head().items():
    for prod in v.items():
        prodMaisLucrativos.append(prod[0])

df_prodMaisLucrativos = df.query(f'Item in {prodMaisLucrativos}')
df_prodMaisLucrativos[['Item ID', 'Item']].groupby(by='Item ID').describe()

Item                                        
        count unique                         top freq
Item ID                                              
30          5      1                 Stormcaller    5
32          6      1                     Orenmir    6
34          9      1             Retribution Axe    9
92          8      1                Final Critic    8
101         6      1                Final Critic    6
115         7      1  Spectral Diamond Doomblade    7
180         5      1                 Stormcaller    5

- O DataFrame acima mostra os produtos mais lucrativos, com o número de vendas e o Item ID.
- Observa-se mais uma vez os dois IDs do produto Final Critic

In [180]:
df_prodMaisLucrativos[['Item', 'Valor']].groupby(by='Valor').describe()

Item                                        
      count unique                         top freq
Valor                                              
1.36      8      1                Final Critic    8
2.78      5      1                 Stormcaller    5
4.14      9      1             Retribution Axe    9
4.15      5      1                 Stormcaller    5
4.25      7      1  Spectral Diamond Doomblade    7
4.62      6      1                Final Critic    6
4.95      6      1                     Orenmir    6

- O DataFrame acima mostra o valor dos produtos mais lucrativos
- Observa-se mais uma vez os dois valores dos produtos Final Critic e Stormcaller

In [183]:
df_prodMaisLucrativos[['Item', 'Valor']].groupby(by='Item').sum().sort_values(by='Valor', ascending=False)

Valor
Item                             
Final Critic                38.60
Retribution Axe             37.26
Stormcaller                 34.65
Spectral Diamond Doomblade  29.75
Orenmir                     29.70

- A tabela acima mostra o valor total vendido dos produtos mais lucrativos.

## Fim

### Obrigado - Data Science Academy - <a href="http://facebook.com/dsacademybr">facebook.com/dsacademybr</a>